src/algebra/ordered_ring.lean

```diff
@@ -3,10 +3,87 @@ Copyright (c) 2016 Jeremy Avigad. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Jeremy Avigad, Leonardo de Moura, Mario Carneiro
 -/
-import algebra.ordered_group
 import algebra.invertible
+import algebra.ordered_group
 import data.set.intervals.basic
 
+/-!
+# Ordered rings and semirings
+
+This file develops the basics of ordered (semi)rings.
+
+Each typeclass here comprises
+* an algebraic class (`semiring`, `comm_semiring`, `ring`, `comm_ring`)
+* an order class (`partial_order`, `linear_order`)
+* assumptions on how both interact ((strict) monotonicity, canonicity)
+
+For short,
+* "`+` respects `≤`" means "monotonicity of addition"
+* "`*` respects `<`" means "strict monotonicity of multiplication by a positive number".
+
+## Typeclasses
+
+* `ordered_semiring`: Semiring with a partial order such that `+` respects `≤` and `*` respects `<`.
+* `ordered_comm_semiring`: Commutative semiring with a partial order such that `+` respects `≤` and
+  `*` respects `<`.
+* `ordered_ring`: Ring with a partial order such that `+` respects `≤` and `*` respects `<`.
+* `ordered_comm_ring`: Commutative ring with a partial order such that `+` respects `≤` and
+  `*` respects `<`.
+* `linear_ordered_semiring`: Semiring with a total order such that `+` respects `≤` and
+  `*` respects `<`.
+* `linear_ordered_ring`: Ring with a total order such that `+` respects `≤` and `*` respects `<`.
+* `linear_ordered_comm_ring`: Commutative ring with a total order such that `+` respects `≤` and
+  `*` respects `<`.
+* `canonically_ordered_comm_semiring`: Commutative semiring with a partial order such that `+` respects `≤`, `*` respects `<`, and `a ≤ b ↔ ∃ c, b = a + c`.
+
+and some typeclasses to define ordered rings by specifying their nonegative elements:
+* `nonneg_ring`: To define `ordered_ring`s.
+* `linear_nonneg_ring`: To define `linear_ordered_ring`s.
+
+## Hierarchy
+
+The hardest part of proving order lemmas might be to figure out the correct generality and its
+corresponding typeclass. Here's an attempt at demystifying it. For each typeclass, we list its
+immediate predecessors and what conditions are added to each of them.
+
+* `ordered_semiring`
+  - `ordered_cancel_add_comm_monoid`: Multiplication. `*` respects `<`.
```

Comment:
To me the `:` made me think that the RHS was explaining the LHS before the `:`.
What do you think of formatting it with `&`'s?
```suggestion
  - `ordered_cancel_add_comm_monoid` & `semiring` & `*` respects `<`.
```

Reply:
It's quite weird but sure.